### PR TITLE
feat(MCP): add Snyk Learn tool [IDE-1213]

### PR DESCRIPTION
**Pull request by rrama**
# 🚀 MCP Learn Tools Integration + Code Quality Improvements

## 📋 Overview
This PR introduces MCP (Model Context Protocol) integration with Snyk Learn tools, along with significant code quality improvements, UI enhancements for ignore functionality, and comprehensive testing additions. The changes span multiple components to provide a cohesive developer experience.

### 🔧 MCP Learn Tools Integration
- **`snyk_open_learn_lesson`** - Opens relevant Snyk Learn lessons in browser based on vulnerability context
- **Smart parameter handling** - Supports CVEs, CWEs, rules, ecosystems, and issue types  
- **Ecosystem mapping** - Automatic translation between package managers and learn ecosystems
- **URL enhancement** - Adds IDE context parameters for better lesson targeting
- **Factory pattern implementation** - Clean dependency injection for learn service

### 📋 Development Experience Enhancements
- **Cursor Rules file** - Complete development guidelines for the project
- **Type safety improvements** - Added `OpenBrowserFunc` type definition

### Learn Service Improvements
- **Fixed `updateCaches()` return value** - Now returns filtered lessons instead of `nil`
- **Enhanced lesson filtering** - Removes unsupported lessons based on description text
- **Resolved empty lesson list bug** - `GetAllLessons()` now returns actual results
- **Simplified ecosystem handling** - Removed redundant npm->javascript mapping

### Code Quality & Testing
- **81.9% test coverage** across MCP extension package
- **560+ lines of comprehensive tests** - Edge cases, error scenarios, mock integration
- **Linting compliance** - Fixed HTTP method constants, empty branches, unused variables
- **Reduced cyclomatic complexity** - Extracted helper functions for maintainability

### Checklist

- [x] Tests added and all succeed
- [ ] Regenerated mocks, etc. (`make generate`)
 - N/A
- [x] Linted (`make lint-fix`)
- [x] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced
 - N/A

[IDE-1213]: https://snyksec.atlassian.net/browse/IDE-1213?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mcp_extension/learn_tool.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+			return nil, err
+		}
+
+		if lessonTitle == "" {
+			err := fmt.Errorf("'lessonTitle' parameter cannot be empty for tool %s", toolDef.Name)
+			logger.Error().Err(err).Msg("Parameter value error")
+			return nil, err
+		}
+
+		logger.Info().Str("lessonTitle", lessonTitle).Msg("Parsed parameters for snyk_open_learn_lesson")
```
Comment (bastiandoetsch) [inline mcp_extension/learn_tool.go:128]: does this need to be an info log? why?

File: mcp_extension/learn_tool.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+		lessonURL := targetLesson.Url
+		if !strings.Contains(lessonURL, "loc=ide") {
+			if strings.Contains(lessonURL, "?") {
+				lessonURL += "&loc=ide"
+			} else {
+				lessonURL += "?loc=ide"
+			}
+		}
+
+		logger.Info().Str("lessonURL", lessonURL).Msg("Attempting to open lesson URL in browser")
```
Comment (bastiandoetsch) [inline mcp_extension/learn_tool.go:166]: this shouldn't be info level

File: mcp_extension/learn_tool_test.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mcp_extension
+
+import (
+	"testing"
+	// Core Go testing package
```
Comment (bastiandoetsch) [inline mcp_extension/learn_tool_test.go:21]: well, these comments are really useless. AI generated?

File: mcp_extension/learn_tool_test.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+	"github.com/mark3labs/mcp-go/mcp"
+
+	"strings" // For URL assertions
+
+	"github.com/snyk/snyk-ls/infrastructure/learn"
+)
+
+// learnTestFixture defines the test fixture.
+// learnTestFixture defines the test fixture.
```
Comment (bastiandoetsch) [inline mcp_extension/learn_tool_test.go:56]: duplicate line

File: mcp_extension/learn_tool_test.go [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+	// Mock Learn Service
+	mockLearnSvc := mock_learn.NewMockService(mockCtrl)
+
+	// Define a test-specific factory that returns the mockLearnSvc
+	testLearnServiceFactory := func(invCtx workflow.InvocationContext, testLogger *zerolog.Logger) learn.Service {
+		return mockLearnSvc
+	}
+
+	// Initialize McpLLMBinding (as done in scan_tool_test.go for its fixture)
+	// We don't need a real CLI path for these tests as learn tools don't directly call CLI.
```
Comment (bastiandoetsch) [inline mcp_extension/learn_tool_test.go:105]: the comments are a bit weird

File: mcp_extension/learn_tool.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+		args := request.GetArguments()
+		cveArray := parseStringArrayParam(args, "cves")
+		cweArray := parseStringArrayParam(args, "cwes")
+		ruleString := parseStringParam(args, "rule")
+		ecosystemString := parseStringParam(args, "ecosystem")
+		issueTypeString := parseStringParam(args, "issueType")
+		issueType := parseIssueType(issueTypeString)
+
+		targetLesson, err := m.learnService.GetLesson(ecosystemString, ruleString, cweArray, cveArray, issueType)
```
Comment (ShawkyZ) [inline mcp_extension/learn_tool.go:110]: check if  m.learnService != nil

**Comment by ShawkyZ** [comment]
Should we wait on merging this change until the custom output PR is implemented ?

File: mcp_extension/learn_tool.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+// buildLessonURL constructs the lesson URL with IDE location parameter
+func buildLessonURL(lessonURL string) (string, error) {
+	u, err := url.Parse(lessonURL)
+	if err != nil {
+		return "", fmt.Errorf("invalid lesson URL: %w", err)
+	}
+	q := u.Query()
+	q.Set("loc", "ide")
```
[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply mcp_extension/learn_tool.go:92]: I've reached out to the team to verify.

**Review by ShawkyZ — APPROVED** [review]
